Address binary serialisation (#12)
input-output-hk/implementation-decisions: text/0000-address.md

proposing a much simpler address format in the sense that it is extendable
within reason.

1. the _compact-ability_ of the address on the blockchain is a prime motivation
   of this proposal. Because of the model of the blockchain, every byte ever
   included in the ledger will be stored there for-ever. We need to ensure that
   we store only what is necessary. On the previous address encoding, the cbor
   model allowed for including unbounded byte array in the address which has been
   leverage to include random data;
2. Another motivation to keep the address as short as possible is that it will
   save on transaction fees too (the current fee algorithm being linear in number
   of bytes);
3. The new address model allows to up to 123 different types of addresses. The
   current RFC only proposes 2 models: simple and grouped.
4. we also propose a new human readable encoding. This will allow visual
   discrimination for the user between the old address format and the new one.

# Guide-level explanation
[guide-level-explanation]: #guide-level-explanation

An `Address` is a mean to associate a token to a cryptographic key pair.

> In order for Bod to send a token to Alice, Alice needs to generate a key

Reviewer: `Bod` -> `Bob`.

Author: thanks

* `ca` is the visual discriminant for **Cardano** address (production address);

So to create a **Single Address** we need a public key.

```
PublicKey = 0102030405060708090a0b0c0d0e0f101112131415161718191a1b1c1d1e1f20
```

A production **Single Address** would look like:
`ca1qvqsyqcyq5rqwzqfpg9scrgwpugpzysnzs23v9ccrydpk8qarc0jqxuzx4s`. Note the `ca` at
the beginning of the `Address`. This is our visual discriminant.

To create a **Grouped Address** we need an additional grouping key:

```rust
GroupKey = 292a2b2c2d2e2f303132333435363738393a3b3c3d3e3f404142434445464748
```

A production **Grouped Address** would look like:
`ca1qsqsyqcyq5rqwzqfpg9scrgwpugpzysnzs23v9ccrydpk8qarc0jq2f29vkz6t30xqcnyve5x5mrwwpe8ganc0f78aqyzsjrg3z5v36gguhxny`.
Note the `ca` prefix again.

A test **Grouped Address** would look like:
`ta1ss5j52ev95hz7vp3xgengdfkxuurjw3m8s7nu06qg9pyx3z9ger5sqgzqvzq2ps8pqys5zcvp58q7yq3zgf3g9gkzuvpjxsmrsw3u8eqx5x7xh`.
Note the `ta` prefix here.

# Reference-level explanation
[reference-level-explanation]: #reference-level-explanation

## Binary serialization

It uses a simple serialization format which is made to be concise:

> * First byte contains the discrimination information (1 bit) and the kind of address (7 bits)

Reviewer: Is the discriminant the most or least significant bit?

Author: It is specified again in the example below but can be specified here again to avoid confusion.

|discriminant | first byte | public key |
|---:|:---|:--|
| Production | `0x03` (`0b0000_0011`) | `0102030405060708090a0b0c0d0e0f101112131415161718191a1b1c1d1e1f20` |
| Testing    | `0x83` (`0b1000_0011`) | `0102030405060708090a0b0c0d0e0f101112131415161718191a1b1c1d1e1f20` |

* Remaining bytes contains a kind specific encoding describe after.

2 kind of address are currently supported:

* Single: Just a (spending) public key using the ED25519 algorithm
* Group: Same as single, but with a added (staking/group) public key
  using the ED25519 algorithm.

```
Single key:
    DISCRIMINATION_BIT || SINGLE_KIND_TYPE (7 bits) || SPENDING_KEY

Group key:
    DISCRIMINATION_BIT || GROUP_KIND_TYPE (7 bits)|| SPENDING_KEY || STAKING_KEY
```

Bit value for the `DISCRIMINATION_BIT`:

| value | purpose |
|---:|:--- |
| 0 | Production Address |

> | 1 | Test Address |

Reviewer: Is two networks enough? E.g. for Cardano alone we already have `mainnet`, `testing` and `staging`, and you might want to prevent address reuse between alternative Cardano-based chains as well...

Reviewer: Ah, I see you addressed this elsewhere.

- Feature Name: address
- Start Date: 2019-01-17
- RFC PR: (leave this empty)
- Authors:
  - Nicolas Di Prima (nicolas.diprima@iohk.io)
  - Vincent Hanquez (vincent.hanquez@iohk.io)

# Summary
[summary]: #summary

A new address format for the cardano blockchain that exposes extendable
properties as well as being as compact as possible.

> The new proposed address format exposes the following properties:

Reviewer 2: ```suggestion
The new proposed address format satisfies the following properties:
```

1. compact on the blockchain;
2. extendable within reason;
3. human readable encoding
  * that includes visual discriminant;
  * with checksum to detect errors;

# Motivation
[motivation]: #motivation

The current address model is wasting quite a lot of space. It is a large

> CBOR structure and does not distinguish what is necessary for the blockchain

Reviewer 3: ```suggestion
CBOR structure and does not distinguish between what is necessary for the blockchain
```

and what is necessary for the human readability. This RFC fixes this by
proposing a much simpler address format in the sense that it is extendable
within reason.

> 1. the _compact-ability_ of the address on the blockchain is a prime motivation

Reviewer 2: ```suggestion
1. The _compact-ability_ of the address on the blockchain is a prime motivation
```

   of this proposal. Because of the model of the blockchain, every byte ever

>    included in the ledger will be stored there for-ever. We need to ensure that

Reviewer 3: ```suggestion
   included in the ledger will be stored there forever. We need to ensure that
```

   we store only what is necessary. On the previous address encoding, the cbor

>    model allowed for including unbounded byte array in the address which has been

Reviewer 3: ```suggestion
   model allowed for including an unbounded byte array in the address which has been
```

>    leverage to include random data;

Reviewer 2: ```suggestion
   leveraged to include random data.
```

Reviewer 2: Or maybe `abused`?

2. Another motivation to keep the address as short as possible is that it will

>    save on transaction fees too (the current fee algorithm being linear in number

Reviewer 2: ```suggestion
   save on transaction fees (the current fee algorithm being linear in number
```
No need for the 'too', you already have "another"

>    of bytes);

Reviewer 2: ```suggestion
   of bytes).
```

3. The new address model allows to up to 123 different types of addresses. The
   current RFC only proposes 2 models: simple and grouped.

> 4. we also propose a new human readable encoding. This will allow visual

Reviewer 2: ```suggestion
4. We also propose a new human readable encoding. This will allow visual
```

   discrimination for the user between the old address format and the new one.

# Guide-level explanation
[guide-level-explanation]: #guide-level-explanation

An `Address` is a mean to associate a token to a cryptographic key pair.
In order for Bob to send a token to Alice, Alice needs to generate a key
pair, a `PublicKey` and a `PrivateKey`. Alice generates an `Address` from
this `PublicKey` and sends the `Address` to Bob and Bob adds on the Ledger
that he has sent a token to the public key.

To create a new address, we only need the following information:

* `PublicKey`: this will be necessary to be able to utilise the token;
* `Discriminant`: this will prevent misuse of addresses between testing
  environment and production environment;

This is the most simple address. It is called a **Single Address**. This is
because it is not associated with any staking key.

It is possible to create a **Grouped Address**, it groups addresses to a
staking key. To create a **Grouped Address** you need the `PublicKey`,
the `Discriminant` (just like for a **Single Address**) and another

> `PublicKey` that is associating this address to another hierarchy.

Reviewer 4: I think we should expand on this, to get the connection to what's in the design document. There, we have four new types of addresses (base, pointer, enterprise, and reward account). This RFC introduces two new kinds of addresses, which can be used to represent three of the addresses in the design document:

- base addresses can be implemented in terms of grouped addresses
- pointer addresses are not covered here
- enterprise addresses can be implemented in terms of single addresses
- reward account addresses can be implemented in terms of grouped addresses

I suggest using the address kind to distinguish the two kinds of grouped addresses (base and reward account), and adding pointer addresses.

Reviewer 5: we're not going to support pointer address. they are not useful for MVP, can be bolted afterwards without a problem from the format PoV, and are raising *important* security questions.

Reviewer 4: So are you saying that with this proposed address format, where the additional staking key hash takes only 32 bytes, the savings of space by using pointer addresses is not enough to warrant the additional complexity?

Reviewer 6: @kantp can you share a document where types of addresses are defined - as I am not aware of the recent development

Reviewer 4: @akegalj They are defined in the [design document for delegation](https://github.com/input-output-hk/fm-ledger-rules/tree/master/docs/delegation_design_spec).

```
PublicKey = 0102030405060708090a0b0c0d0e0f101112131415161718191a1b1c1d1e1f20
```

A production **Single Address** would look like:
`ca1qvqsyqcyq5rqwzqfpg9scrgwpugpzysnzs23v9ccrydpk8qarc0jqxuzx4s`. Note the `ca` at
the beginning of the `Address`. This is our visual discriminant.

To create a **Grouped Address** we need an additional grouping key:

```rust
GroupKey = 292a2b2c2d2e2f303132333435363738393a3b3c3d3e3f404142434445464748
```

A production **Grouped Address** would look like:
`ca1qsqsyqcyq5rqwzqfpg9scrgwpugpzysnzs23v9ccrydpk8qarc0jq2f29vkz6t30xqcnyve5x5mrwwpe8ganc0f78aqyzsjrg3z5v36gguhxny`.
Note the `ca` prefix again.

A test **Grouped Address** would look like:
`ta1ss5j52ev95hz7vp3xgengdfkxuurjw3m8s7nu06qg9pyx3z9ger5sqgzqvzq2ps8pqys5zcvp58q7yq3zgf3g9gkzuvpjxsmrsw3u8eqx5x7xh`.
Note the `ta` prefix here.

# Reference-level explanation
[reference-level-explanation]: #reference-level-explanation

## Binary serialization

It uses a simple serialization format which is made to be concise:

* First byte contains the discrimination information (1 bit) and the kind of address (7 bits).
  The discriminant being the most significant bit.

> * Remaining bytes contains a kind specific encoding describe after.

Reviewer 2: ```suggestion
* Remaining bytes contains a kind specific encoding described after.
```

> 2 kind of address are currently supported:

Reviewer 2: ```suggestion
2 kinds of address are currently supported:
```

See [bech32].

# Drawbacks
[drawbacks]: #drawbacks

N/A

# Rationale and alternatives
[rationale-and-alternatives]: #rationale-and-alternatives

Current addresses are in average about 74 bytes long (and this is without the
stake key).

* a hash with a given 28 bytes size (`Blake2b-224`),
* a HDPayload of variable size, containing a:
  * 16 bytes poly1305 authentication tag
  * an encrypted CBOR encoded array of 32 bits derivation indices
* stake distribution, either:
  * nothing when using bootstrap era
  * a single key distribution, a hash with a given 32bytes size (`Blake2b-256`)
* Address type Tag
* 4 bytes of crc32
* 4 bytes of address discrimination

On the blockchain there is no reason to store the crc32. This is only
useful for the human readable encoding part to prevent mistyping.
There is also no need to have more than 1 bit of address discrimination
on the blockchain: the signature protocol takes into account the
protocol magic (the blockchain discriminant).

The new format only contains:

* 1 bytes of metadata:
  * 1bit of address discrimination (production/testing);
  * 7bits of address type (allowing up to 123 different types);
* 32 bytes of public key;

> * 32 bytes of stake distribution (for non bootstrap era)

Reviewer 4: ```suggestion
* 32 bytes of public staking key (optional)
```
